#!/bin/bash

#______________________________________________________________________________________________________________
###############################################################################################################
#                                                                                                             #
# This script calls variants in BAM files generated by bwa mem and sorted and indexed by samtools using gatk4  #
#                                                                                                             #
###############################################################################################################

PATH_IN_BAMS=/media/rna/INIA/benchmark/Alignment/BWA/BWA_BAM
PATH_REF_FILE=/media/rna/INIA/benchmark/Variant_calling/gatk/Reference/DM_1-3_516_R44_potato_genome_assembly.v6.1.fa
PATH_VCF_OUT=/media/rna/INIA/benchmark/Variant_calling/gatk/bwa/VCF_output
mkdir -p $PATH_VCF_OUT
PATH_STAT_OUT=/media/rna/INIA/benchmark/Variant_calling/gatk/bwa/stat
mkdir -p $PATH_STAT_OUT

# Calculating number of total input samples
TOTAL_SAMPLES=$(ls -l $PATH_IN_BAMS | grep -c ^d)


#create sequence dictionary as gatk needs a fasta dictionary for the reference
#/media/rna/INIA/benchmark/Variant_calling/gatk/gatk-4.2.5.0/gatk CreateSequenceDictionary -R $PATH_REF_FILE
#/media/rna/INIA/benchmark/Variant_calling/gatk/gatk-4.2.5.0/gatk CreateSequenceDictionary -R DM_1-3_516_R44_potato_genome_assembly.v6.1.fa
#samtools faidx DM_1-3_516_R44_potato_genome_assembly.v6.1.fa

#save the terminal output in a txt file
#script /media/rna/INIA/benchmark/Variant_calling/gatk/bwa/script.txt 

for folder in ${PATH_IN_BAMS}/*; do
	sample=$(basename $folder)

    # Output path and name for the VCF file for each sample
    VCF_OUT_FILE=$PATH_VCF_OUT/${sample}.vcf
    # Specifying file name for compuation times for each sample
    COMP=${PATH_STAT_OUT}/computation_times_${sample}.txt

    gatk_start=`date +%s`
    
   

    #validate and check for the read groups
    /media/rna/INIA/benchmark/Variant_calling/gatk/gatk-4.2.5.0/gatk ValidateSamFile   -I $PATH_IN_BAMS/$sample/*_sorted*.bam -R $PATH_REF_FILE MODE= SUMMARY
    
    #fix the readgroup issue
    /media/rna/INIA/benchmark/Variant_calling/gatk/gatk-4.2.5.0/gatk AddOrReplaceReadGroups -I $PATH_IN_BAMS/$sample/*_sorted*.bam -O $PATH_VCF_OUT/$sample.bam -SORT_ORDER coordinate -RGID foo -RGLB bar -RGPL illumina -RGPU barcode  -RGSM Sample1 -CREATE_INDEX True
    
    #variant calling
    /media/rna/INIA/benchmark/Variant_calling/gatk/gatk-4.2.5.0/gatk HaplotypeCaller -R $PATH_REF_FILE -I $PATH_VCF_OUT/$sample.bam -ploidy 4 -O $VCF_OUT_FILE
    gatk_end=`date +%s`
    gatk_time=$((gatk_end-gatk_start))
    echo "gatk variant calling time in s ($sample):	$gatk_time" >> $COMP

done 

exit
